### PR TITLE
refactor: Remove support for protocol version 2.1

### PR DESCRIPTION
**Pull request by bthomee**
## High Level Overview of Change

This change removes support for protocol version 2.1.

### Context of Change

Protocol version 2.1 was introduced years ago, and has no longer been in use since https://github.com/XRPLF/rippled/commit/4b9d3ca7, which introduced version 2.2.

Similar to retiring amendments that have been enabled for more than 2 years, old protocol versions and the features that they introduced can also be retired when they have not been supported for at least that amount of time.

This PR essentially permanently enables protocol feature `ValidatorList2Propagation` (which requires protocol version 2.2, introduced 5 years ago) and removes support for protocol feature `ValidatorListPropagation`. The `2` is kept in variable and constant names for simplicity. Also, in case we ever need to introduce a new protocol feature/version to support another change in the validator list handling, it's useful to keep the 2 there.

### CODE REVIEW

File: src/xrpld/app/misc/detail/ValidatorList.cpp [hunk 1]
```diff
@@ -897,7 +879,7 @@ ValidatorList::broadcastBlobs(
                         lists.rawVersion,
                         lists.rawManifest,
                         blobInfos,
-                        v2 ? messages2[peerSequence] : messages1,
+                        messages2[peerSequence],
                         hashRouter,
                         j);
                     // Even if the peer doesn't support the messages,
```
[the remark this answers is not in the document]
Comment (bthomee) [reply src/xrpld/app/misc/detail/ValidatorList.cpp]: Done.

File: src/xrpld/app/misc/detail/ValidatorList.cpp [hunk 2]
```diff
@@ -727,17 +727,8 @@ ValidatorList::sendValidatorList(
...
-    else if (peer.supportsFeature(ProtocolFeature::ValidatorListPropagation))
-    {
-        messageVersion = 1;
-    }
-    if (messageVersion == 0u)
-        return;
+    // v1 messages are no longer supported.
+    std::size_t const messageVersion = 2;
     auto const [newPeerSequence, numVLs] = buildValidatorListMessages(
         messageVersion, peerSequence, maxSequence, rawVersion, rawManifest, blobInfos, messages);
```
[the remark this answers is not in the document]
Comment (bthomee) [reply src/xrpld/app/misc/detail/ValidatorList.cpp]: Done.

File: src/xrpld/overlay/detail/PeerImp.cpp [hunk 3]
```diff
@@ -2275,14 +2271,6 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMValidatorList> const& m)
...
     {
-        if (!supportsFeature(ProtocolFeature::ValidatorListPropagation))
-        {
-            JLOG(pJournal_.debug()) << "ValidatorList: received validator list from peer using "
-                                    << "protocol version " << to_string(protocol_)
-                                    << " which shouldn't support this feature.";
-            fee_.update(Resource::kFeeUselessData, "unsupported peer");
-            return;
-        }
         onValidatorListMessage(
```
[the remark this answers is not in the document]
Comment (bthomee) [reply src/xrpld/overlay/detail/PeerImp.cpp]: Done.

File: src/test/app/ValidatorList_test.cpp [hunk 4]
```diff
@@ -2456,68 +2401,34 @@ class ValidatorList_test : public beast::unit_test::Suite
...
-        }
-
-        // Version 2
+        // Computes the exact wire size of a TMValidatorListCollection
+        // message containing the given sequences, so tests below can assert
+        // on message sizes without hardcoding numbers that would need to be
+        // hand-recalculated (and re-verified against the real serializer)
+        // every time the message format changes.
+        auto const messageSize = [&manifest,
+                                  &blobInfos](std::vector<std::uint32_t> const& sequences) {
```
[the remark this answers is not in the document]
Comment (bthomee) [reply src/test/app/ValidatorList_test.cpp]: Clang-tidy flagged it as not needed. Namely, `version` was captured by the lambda but only ever used inside via `msg.set_version(version)`; clang-tidy correctly flagged that since version is a `std::uint32_t const` initialized to the compile-time literal `2`, so the compiler treats its use as a constant expression rather than requiring capture, making the explicit capture unused.

File: src/xrpld/app/misc/detail/ValidatorList.cpp [hunk 5]
```diff
@@ -556,51 +526,34 @@ splitMessageParts(
...
     }
 
+    // An unsplittable blob is still bounded by the protocol limit: peers drop
+    // messages exceeding it on receipt, so don't waste the bandwidth. maxSize
+    // only ever tightens this (it defaults to kMaximumMessageSize), so a blob
+    // reaching here can exceed maxSize but never the protocol limit.
+    if (size > kMaximumMessageSize)
+    {
+        UNREACHABLE("xrpl::splitMessageParts : maximum message size exceeded");
+        return 0;
```
Comment (ximinez) [inline src/xrpld/app/misc/detail/ValidatorList.cpp:548]: You can add coverage exclusions here

File: src/xrpld/overlay/detail/PeerImp.cpp [hunk 6]
```diff
@@ -541,10 +541,6 @@ PeerImp::supportsFeature(ProtocolFeature f) const
 {
     switch (f)
     {
-        case ProtocolFeature::ValidatorListPropagation:
-            return protocol_ >= makeProtocol(2, 1);
-        case ProtocolFeature::ValidatorList2Propagation:
-            return protocol_ >= makeProtocol(2, 2);
```
Comment (ximinez) [inline src/xrpld/overlay/detail/PeerImp.cpp:548]: For my own reference: Now that the minimum supported version is 2.2, any peer that successfully connects will be guaranteed to have a protocol version >= 2.2, so that's why this check can be removed. Arguably, the v2.1 check could have been removed when versions old than 2.1 were removed, and for the same reason. It was probably just missed at that time.

File: src/test/app/ValidatorList_test.cpp [hunk 7]
```diff
@@ -2334,92 +2310,81 @@ class ValidatorList_test : public beast::unit_test::Suite
...
-                std::vector<std::pair<std::size_t, std::vector<std::uint32_t>>> expectedInfo) {
-                BEAST_EXPECT(messages.size() == expectedInfo.size());
-                auto msgIter = expectedInfo.begin();
-                for (auto const& messageWithHash : messages)
+        // Computes the exact wire size of a TMValidatorListCollection
+        // message containing the given sequences, so verifyMessage below can
+        // assert on message sizes without hardcoding numbers that would
+        // need to be hand-recalculated (and re-verified against the real
+        // serializer) every time the message format changes.
+        auto const messageSize = [](auto const version,
```
Comment (ximinez) [inline src/test/app/ValidatorList_test.cpp]: How often is the message format going to change, especially for validator lists? Wouldn't we want to verify with known values that, absent any expect changes, the format and size of the message _doesn't_ change? If a regression or incompatibility is unintentionally introduced into the serializer, would this test detect that?

Also, `messageWithHash.message` below is a `std::shared_ptr<Message>`, which is built from a `TMValidatorListCollection` in `buildValidatorListMessage` with the same inputs as this lambda, that does the same thing as this lamba, so the check that `size == messageSize(...)` is pretty much a tautology. The only way it wouldn't be true is if one function changed (due to a message format change, or a mistake), but not the other. That would alert the dev, true, but wouldn't tell them if they broke the message, or just need to update this function, too, thus turning it back into a tautology.

File: src/test/overlay/ProtocolVersion_test.cpp [hunk 8]
```diff
@@ -36,7 +36,6 @@ class ProtocolVersion_test : public beast::unit_test::Suite
         testcase("Convert protocol version to string");
         BEAST_EXPECT(to_string(makeProtocol(1, 3)) == "XRPL/1.3");
         BEAST_EXPECT(to_string(makeProtocol(2, 0)) == "XRPL/2.0");
-        BEAST_EXPECT(to_string(makeProtocol(2, 1)) == "XRPL/2.1");
```
Comment (ximinez) [inline src/test/overlay/ProtocolVersion_test.cpp:39]: Strictly speaking, you don't need to remove this test, since it only tests how the generation and stringification work. Note that there are even older versions, and a non-existent future version tested here as well.

File: src/test/overlay/ProtocolVersion_test.cpp
Comment (ximinez) [inline src/test/overlay/ProtocolVersion_test.cpp:79]: For completeness here, could you create another test case that checks that the highest version is returned if multiple valid versions are provided? I just realized that, even before your change, None of these tests included two valid versions.

```
            BEAST_EXPECT(
                negotiateProtocolVersion("XRPL/2.1 XRPL/2.2, XRPL/2.3, XRPL/2.4, XRPL/999.999") ==
                makeProtocol(2, 3));
```

**Review by ximinez — CHANGES_REQUESTED** [review]

File: include/xrpl/proto/xrpl.proto
Comment (pratikmankawde) [inline include/xrpl/proto/xrpl.proto:28]: I was curious about the gaps and the message types shipped against those. So, I went through the histroy of this file, across several renames, and found that some numbers were never used. Would it be worth putting this table in the comments for curious people like me?
| #     | Message type(s) ever assigned                           | First seen | Last seen  | Status today             |
| ----- | ------------------------------------------------------- | ---------- | ---------- | ------------------------ |
| 0     | mtHELLO                                                 | 2011-12-19 | 2011-12-19 | retired                  |
| 1     | mtERROR_MSG → mtHELLO                                   | 2011-12-19 | 2019-04-10 | retired                  |
| 2     | mtPING → mtERROR_MSG → mtMANIFESTS                      | 2011-12-19 | current    | live                     |
| 3     | mtPING                                                  | 2012-01-19 | current    | live                     |
| 4     | mtPOW → mtPROOFOFWORK                                   | 2012-11-21 | 2019-04-10 | retired                  |
| 5     | mtCLUSTER                                               | 2013-07-09 | current    | live                     |
| 6     | —                                                       | —          | —          | never used               |
| 7     | —                                                       | —          | —          | never used               |
| 8     | —                                                       | —          | —          | never used               |
| 9     | —                                                       | —          | —          | never used               |
| 10    | mtGET_CONTACTS                                          | 2011-12-19 | 2014-05-01 | retired                  |
| 11    | mtCONTACT                                               | 2011-12-19 | 2014-05-01 | retired                  |
| 12    | mtGET_PEERS                                             | 2012-06-13 | 2019-04-10 | retired                  |
| 13    | mtPEERS                                                 | 2012-06-13 | 2019-04-10 | retired                  |
| 14    | mtANNOUNCE → mtENDPOINTS → mtUNUSED_FIELD               | 2013-09-11 | 2014-05-01 | retired                  |
| 15    | mtENDPOINTS                                             | 2014-02-05 | current    | live                     |
| 16    | —                                                       | —          | —          | never used               |
| 17    | —                                                       | —          | —          | never used               |
| 18    | —                                                       | —          | —          | never used               |
| 19    | —                                                       | —          | —          | never used               |
| 20    | mtSEARCH_TRANSACTION                                    | 2011-12-19 | 2014-05-01 | retired                  |
| 21    | mtGET_ACCOUNT                                           | 2011-12-19 | 2014-05-01 | retired                  |
| 22    | mtACCOUNT                                               | 2011-12-19 | 2014-05-01 | retired                  |
| 23–29 | —                                                       | —          | —          | never used               |
| 30    | mtTRANSACTION                                           | 2011-12-19 | current    | live                     |
| 31    | mtGET_LEDGER                                            | 2011-12-19 | current    | live                     |
| 32    | mtLEDGER → mtLEDGER_DATA                                | 2011-12-19 | current    | live                     |
| 33    | mtPROPOSE_LEDGER                                        | 2011-12-19 | current    | live                     |
| 34    | mtSTATUS_CHANGE                                         | 2012-05-29 | current    | live                     |
| 35    | mtCLOSE_LEDGER → mtHAVE_SET                             | 2011-12-19 | current    | live                     |
| 36    | mtSTATUS_CHANGE                                         | 2012-05-08 | 2012-05-28 | retired (moved to 34)    |
| 37    | mtPEER_POSITION                                         | 2012-05-15 | 2012-05-28 | retired                  |
| 38    | mtHAVE_SET                                              | 2012-05-22 | 2012-05-28 | retired (moved to 35)    |
| 39    | —                                                       | —          | —          | never used               |
| 40    | mtGET_VALIDATIONS                                       | 2011-12-19 | 2014-05-01 | retired                  |
| 41    | mtVALIDATION                                            | 2011-12-19 | current    | live                     |
| 42    | mtGET_OBJECT → mtGET_OBJECTS                            | 2011-12-19 | current    | live                     |
| 43    | mtOBJECT                                                | 2011-12-19 | 2012-09-11 | retired                  |
| 44–49 | —                                                       | —          | —          | never used               |
| 50    | mtGET_SHARD_INFO                                        | 2018-07-10 | 2024-06-20 | retired (shards removed) |
| 51    | mtSHARD_INFO                                            | 2018-07-10 | 2024-06-20 | retired (shards removed) |
| 52    | mtGET_PEER_SHARD_INFO                                   | 2019-04-10 | 2024-06-20 | retired (shards removed) |
| 53    | mtPEER_SHARD_INFO                                       | 2019-04-10 | 2024-06-20 | retired (shards removed) |
| 54    | mtVALIDATORLIST → mtVALIDATOR_LIST                      | 2019-04-24 | 2026-07-30 | retired by this PR       |
| 55    | mtSQUELCH                                               | 2020-04-12 | current    | live                     |
| 56    | mtVALIDATORLISTCOLLECTION → mtVALIDATOR_LIST_COLLECTION | 2020-07-20 | current    | live                     |
| 57    | mtPROOF_PATH_REQ                                        | 2020-07-20 | current    | live                     |
| 58    | mtPROOF_PATH_RESPONSE                                   | 2020-07-20 | current    | live                     |
| 59    | mtREPLAY_DELTA_REQ                                      | 2020-07-20 | current    | live                     |
| 60    | mtREPLAY_DELTA_RESPONSE                                 | 2020-07-20 | current    | live                     |
| 61    | mtGET_PEER_SHARD_INFO_V2                                | 2020-09-08 | 2024-06-20 | retired (shards removed) |
| 62    | mtPEER_SHARD_INFO_V2                                    | 2020-09-08 | 2024-06-20 | retired (shards removed) |
| 63    | mtHAVE_TRANSACTIONS                                     | 2020-09-08 | current    | live                     |
| 64    | mtTRANSACTIONS                                          | 2020-09-08 | current    | live                     |
| 65    | mtSTART_PROTOCOL                                        | 2023-09-22 | 2023-11-30 | retired (reverted)       |
| 66    | mtGRACEFUL_CLOSE                                        | 2023-09-22 | 2023-11-30 | retired (reverted)       |

Comment (pratikmankawde) [reply include/xrpl/proto/xrpl.proto:28]: Code friendly:
```
| #     | Message type(s) ever assigned                           | First seen | Last seen  | Status today             |
| ----- | ------------------------------------------------------- | ---------- | ---------- | ------------------------ |
| 0     | mtHELLO                                                 | 2011-12-19 | 2011-12-19 | retired                  |
| 1     | mtERROR_MSG → mtHELLO                                   | 2011-12-19 | 2019-04-10 | retired                  |
| 2     | mtPING → mtERROR_MSG → mtMANIFESTS                      | 2011-12-19 | current    | live                     |
| 3     | mtPING                                                  | 2012-01-19 | current    | live                     |
| 4     | mtPOW → mtPROOFOFWORK                                   | 2012-11-21 | 2019-04-10 | retired                  |
| 5     | mtCLUSTER                                               | 2013-07-09 | current    | live                     |
| 6     | —                                                       | —          | —          | never used               |
| 7     | —                                                       | —          | —          | never used               |
| 8     | —                                                       | —          | —          | never used               |
| 9     | —                                                       | —          | —          | never used               |
| 10    | mtGET_CONTACTS                                          | 2011-12-19 | 2014-05-01 | retired                  |
| 11    | mtCONTACT                                               | 2011-12-19 | 2014-05-01 | retired                  |
| 12    | mtGET_PEERS                                             | 2012-06-13 | 2019-04-10 | retired                  |
| 13    | mtPEERS                                                 | 2012-06-13 | 2019-04-10 | retired                  |
| 14    | mtANNOUNCE → mtENDPOINTS → mtUNUSED_FIELD               | 2013-09-11 | 2014-05-01 | retired                  |
| 15    | mtENDPOINTS                                             | 2014-02-05 | current    | live                     |
| 16    | —                                                       | —          | —          | never used               |
| 17    | —                                                       | —          | —          | never used               |
| 18    | —                                                       | —          | —          | never used               |
| 19    | —                                                       | —          | —          | never used               |
| 20    | mtSEARCH_TRANSACTION                                    | 2011-12-19 | 2014-05-01 | retired                  |
| 21    | mtGET_ACCOUNT                                           | 2011-12-19 | 2014-05-01 | retired                  |
| 22    | mtACCOUNT                                               | 2011-12-19 | 2014-05-01 | retired                  |
| 23–29 | —                                                       | —          | —          | never used               |
| 30    | mtTRANSACTION                                           | 2011-12-19 | current    | live                     |
| 31    | mtGET_LEDGER                                            | 2011-12-19 | current    | live                     |
| 32    | mtLEDGER → mtLEDGER_DATA                                | 2011-12-19 | current    | live                     |
| 33    | mtPROPOSE_LEDGER                                        | 2011-12-19 | current    | live                     |
| 34    | mtSTATUS_CHANGE                                         | 2012-05-29 | current    | live                     |
| 35    | mtCLOSE_LEDGER → mtHAVE_SET                             | 2011-12-19 | current    | live                     |
| 36    | mtSTATUS_CHANGE                                         | 2012-05-08 | 2012-05-28 | retired (moved to 34)    |
| 37    | mtPEER_POSITION                                         | 2012-05-15 | 2012-05-28 | retired                  |
| 38    | mtHAVE_SET                                              | 2012-05-22 | 2012-05-28 | retired (moved to 35)    |
| 39    | —                                                       | —          | —          | never used               |
| 40    | mtGET_VALIDATIONS                                       | 2011-12-19 | 2014-05-01 | retired                  |
| 41    | mtVALIDATION                                            | 2011-12-19 | current    | live                     |
| 42    | mtGET_OBJECT → mtGET_OBJECTS                            | 2011-12-19 | current    | live                     |
| 43    | mtOBJECT                                                | 2011-12-19 | 2012-09-11 | retired                  |
| 44–49 | —                                                       | —          | —          | never used               |
| 50    | mtGET_SHARD_INFO                                        | 2018-07-10 | 2024-06-20 | retired (shards removed) |
| 51    | mtSHARD_INFO                                            | 2018-07-10 | 2024-06-20 | retired (shards removed) |
| 52    | mtGET_PEER_SHARD_INFO                                   | 2019-04-10 | 2024-06-20 | retired (shards removed) |
| 53    | mtPEER_SHARD_INFO                                       | 2019-04-10 | 2024-06-20 | retired (shards removed) |
| 54    | mtVALIDATORLIST → mtVALIDATOR_LIST                      | 2019-04-24 | 2026-07-30 | retired by this PR       |
| 55    | mtSQUELCH                                               | 2020-04-12 | current    | live                     |
| 56    | mtVALIDATORLISTCOLLECTION → mtVALIDATOR_LIST_COLLECTION | 2020-07-20 | current    | live                     |
| 57    | mtPROOF_PATH_REQ                                        | 2020-07-20 | current    | live                     |
| 58    | mtPROOF_PATH_RESPONSE                                   | 2020-07-20 | current    | live                     |
| 59    | mtREPLAY_DELTA_REQ                                      | 2020-07-20 | current    | live                     |
| 60    | mtREPLAY_DELTA_RESPONSE                                 | 2020-07-20 | current    | live                     |
| 61    | mtGET_PEER_SHARD_INFO_V2                                | 2020-09-08 | 2024-06-20 | retired (shards removed) |
| 62    | mtPEER_SHARD_INFO_V2                                    | 2020-09-08 | 2024-06-20 | retired (shards removed) |
| 63    | mtHAVE_TRANSACTIONS                                     | 2020-09-08 | current    | live                     |
| 64    | mtTRANSACTIONS                                          | 2020-09-08 | current    | live                     |
| 65    | mtSTART_PROTOCOL                                        | 2023-09-22 | 2023-11-30 | retired (reverted)       |
| 66    | mtGRACEFUL_CLOSE                                        | 2023-09-22 | 2023-11-30 | retired (reverted)       |

```

File: src/test/overlay/ProtocolVersion_test.cpp (hunk 8 above)
Comment (pratikmankawde) [reply src/test/overlay/ProtocolVersion_test.cpp:39]: An alternate perspective would be, do we need multiple tests testing pretty much the same thing? I would instead recommend to add tests for edge cases.
Like: 
```
makeProtocol(0, 1),
makeProtocol(0, 1.0), 
makeProtocol(0.0, 1), 
makeProtocol(0, 0), 
makeProtocol(65535, 65535), 
makeProtocol(-65536, 65535)
```

File: src/xrpld/overlay/detail/ProtocolMessage.h [hunk 9]
```diff
@@ -404,9 +402,6 @@ invokeProtocolMessage(Buffers const& buffers, Handler& handler, std::size_t& hin
         case protocol::mtVALIDATION:
             success = detail::invoke<protocol::TMValidation>(*header, buffers, handler);
             break;
-        case protocol::mtVALIDATOR_LIST:
-            success = detail::invoke<protocol::TMValidatorList>(*header, buffers, handler);
-            break;
```
Comment (pratikmankawde) [inline src/xrpld/overlay/detail/ProtocolMessage.h:429]: I suppose this means `mtVALIDATOR_LIST` is unsupported message from this point onwards. Earlier when it was supported but was being deprecated, did we penalize users sending this message? Will we penalize user sending this message now? I suppose we don't have any code to smartly handle flood of message types we don't support anymore?
Can the `kFeeUselessData` fee be applied to similar message types?

**Review by pratikmankawde — CHANGES_REQUESTED** [review]
Left a-few comments.

File: include/xrpl/proto/xrpl.proto
Comment (bthomee) [reply include/xrpl/proto/xrpl.proto:28]: I think this would be good for a follow-up, where the fields are removed and the numbers are reserved.

I see you have 6-10 listed as never used, although the comments in the file say they were once used...

File: src/xrpld/overlay/detail/ProtocolMessage.h (hunk 9 above)
Comment (bthomee) [reply src/xrpld/overlay/detail/ProtocolMessage.h:429]: Good questions, and there's some follow-up work for us to do here.
* If you send a message that's recognized but not appropriate for what you're calling, you get charged `kFeeUselessData`.
* If you send a message that's not recognized then nothing happens.

As this PR drops `mtVALIDATOR_LIST`, it means that peers who send it won't be charged anything anymore, although - like the other retired fields - there's no reason for a peer to send such a message.

Arguably, `onMessageUnknown(type)` has the type and can call `fee_.update(resource::kFeeUselessData, "unknown message")` (or charge() directly, since `onMessageEnd` won't be invoked to do it for us). I think we should start charging a fee going forward.

File: src/test/app/ValidatorList_test.cpp (hunk 7 above)
Comment (bthomee) [reply src/test/app/ValidatorList_test.cpp]: I don't think the message format will change often. However, even if it does, checking exact serialized size doesn't validate anything meaningful on its own. Wire-format stability is protobuf's job, not something we need to spot-check per message; if it mattered, it'd need a dedicated golden-byte test at the protocol layer covering every message type; we only checked it for the validation messages but actually for no useful purpose.

I had kept the general gist of the size checks to not deviate too much from what it used to be (before message sizes had been calculated offline and the numbers hardcoded), but have now removed it completely since, as I explained above, doing so is pointless.

If this or any other message changes in the future, the author is expected to update the unit tests accordingly; existing unit test coverage is expected to catch other regressions. That being said, we have coverage gaps that we have been working on to close. For instance, `sendValidatorList`/`broadcastBlobs` aren't called from any test — they're only exercised indirectly through integration-style peer network tests, not `ValidatorList_test.cpp`. This is a pre-existing coverage gap that we need to close in the near future.

File: include/xrpl/proto/xrpl.proto
Comment (pratikmankawde) [reply include/xrpl/proto/xrpl.proto:28]: Couldn't flag those 6-10 in commit history.

File: src/test/overlay/ProtocolVersion_test.cpp
Comment (bthomee) [reply src/test/overlay/ProtocolVersion_test.cpp:79]: Done

File: src/test/overlay/ProtocolVersion_test.cpp (hunk 8 above)
Comment (bthomee) [reply src/test/overlay/ProtocolVersion_test.cpp:39]: Done. Expanded multiple tests, although note that `makeProtocol` only takes unsigned ints; I added `-65536` instead in some of the version strings that were parsed.

File: src/xrpld/app/misc/detail/ValidatorList.cpp (hunk 5 above)
Comment (bthomee) [reply src/xrpld/app/misc/detail/ValidatorList.cpp:548]: Done.

File: src/test/app/ValidatorList_test.cpp (hunk 7 above)
Comment (ximinez) [reply src/test/app/ValidatorList_test.cpp]: > I don't think the message format will change often. However, even if it does, checking exact serialized size doesn't validate anything meaningful on its own. Wire-format stability is protobuf's job, not something we need to spot-check per message; if it mattered, it'd need a dedicated golden-byte test at the protocol layer covering every message type; we only checked it for the validation messages but actually for no useful purpose.
>
> I had kept the general gist of the size checks to not deviate too much from what it used to be (before message sizes had been calculated offline and the numbers hardcoded), but have now removed it completely since, as I explained above, doing so is pointless.

When you put it that way, I agree.

> For instance, `sendValidatorList`/`broadcastBlobs` aren't called from any test — they're only exercised indirectly through integration-style peer network tests, not `ValidatorList_test.cpp`. This is a pre-existing coverage gap that we need to close in the near future.

It's nearly impossible to test most `Peer` functionality right now, just owing to the fact that the test would have to create and manage two peers on a closed, private network.

**Review by ximinez — APPROVED** [review]
Thanks for those updates!

**Review by pratikmankawde — COMMENTED** [review]
Left  a minor comment.

**Review by pratikmankawde — APPROVED** [review]
LGTM.
Please remember to create a ticket for charging fee on receiving unsupported message types.
